skopt/tree_opt.py

```diff
@@ -0,0 +1,256 @@
+          `Categorical`).
+
+    * `base_estimator` [string or `Regressor`, default=`"rf"`]:
+        The regressor to use as surrogate model. Can be either
+
+        - `"rf"` for random forest regressor
+        - `"et"` for extra trees regressor
+        - `"dt"` for single decision tree regressor
+        - instance of regressor with support for `with_std` in its predict
+          method
+
+        The predefined models are initilized with good defaults. If you
+        want to adjust the model parameters pass your own instance of
+        a regressor which returns the mean and standard deviation when
+        making predictions.
+
+    * `maxiter` [int, default=100]:
+        Number of iterations used to find the minimum. This corresponds
+        to the total number of evaluations of `func`. If `n_start` > 0
+        only `maxiter - n_start` iterations are used.
+
+    * `n_start` [int, default=10]:
+        Number of random points to draw before fitting `base_estimator`
+        for the first time. If `n_start > maxiter` this degrades to
+        a random search for the minimum.
+
+    * `n_points` [int, default=1000]:
+        Number of points to sample when minimizing the acquisition function.
+
+    * `random_state` [int, RandomState instance, or None (default)]:
+        Set random state to something other than None for reproducible
+        results.
+
+    Returns
+    -------
+    * `res` [`OptimizeResult`, scipy object]:
+        The optimization result returned as a OptimizeResult object.
+        Important attributes are:
+
+        - `x` [float]: location of the minimum.
+        - `fun` [float]: function value at the minimum.
+        - `models`: surrogate models used for each iteration.
+        - `x_iters` [array]: location of function evaluation for each
+           iteration.
+        - `func_vals` [array]: function value for each iteration.
+
+        For more details related to the OptimizeResult object, refer
+        http://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.OptimizeResult.html
+    """
+    rng = check_random_state(random_state)
+
+    if isinstance(base_estimator, str):
+        if base_estimator not in ("rf", "et", "dt"):
+            raise ValueError("Valid values for the base_estimator parameter"
+                             " are: 'rf', 'et' or 'dt', not '%s'" % base_estimator)
+
+        if base_estimator == "rf":
+            base_estimator = RandomForestRegressor(min_samples_leaf=10,
+                                                   random_state=rng)
+
+        elif base_estimator == "et":
+            base_estimator = ExtraTreesRegressor(min_samples_leaf=10,
+                                                 random_state=rng)
+
+        elif base_estimator == "dt":
+            base_estimator = DecisionTreeRegressor(min_samples_leaf=10,
+                                                   random_state=rng)
+
+    else:
+        if not (hasattr(base_estimator, '_estimator_type') and
```

Comment:
`isinstance(RegressorMixin)` does not allow ducktyping, since it requires the object to inherit from the mixin. The official way, as we describe it in scikit-learn, is to check for that attribute.